src/datasets/packaged_modules/cache/cache.py

```diff
@@ -0,0 +1,153 @@
+import glob
+import os
+import shutil
+import time
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import pyarrow as pa
+
+import datasets
+import datasets.config
+from datasets.naming import filenames_for_dataset_split
+
+
+logger = datasets.utils.logging.get_logger(__name__)
+
+
+def _get_modification_time(cached_directory_path):
+    return (Path(cached_directory_path)).stat().st_mtime
+
+
+def _find_hash_in_cache(
+    dataset_name: str, config_name: Optional[str], version: Optional[str], cache_dir: Optional[str]
+) -> Tuple[str, str, str]:
+    cache_dir = os.path.expanduser(str(cache_dir or datasets.config.HF_DATASETS_CACHE))
+    cached_datasets_directory_path_root = os.path.join(cache_dir, dataset_name.replace("/", "___"))
+    cached_directory_paths = [
+        cached_directory_path
+        for cached_directory_path in glob.glob(
+            os.path.join(cached_datasets_directory_path_root, config_name or "*", version or "*", "*")
+        )
+        if os.path.isdir(cached_directory_path)
+    ]
+    if not cached_directory_paths:
+        if config_name is not None:
+            cached_directory_paths = [
+                cached_directory_path
+                for cached_directory_path in glob.glob(
+                    os.path.join(cached_datasets_directory_path_root, "*", "*", "*")
+                )
+                if os.path.isdir(cached_directory_path)
+            ]
+        available_configs = sorted(
+            {Path(cached_directory_path).parts[-3] for cached_directory_path in cached_directory_paths}
+        )
+        available_versions = sorted(
+            {Path(cached_directory_path).parts[-2] for cached_directory_path in cached_directory_paths}
+        )
+        raise ValueError(
+            f"Couldn't find cache for {dataset_name}"
+            + (f" for config '{config_name}'" if config_name else "")
+            + (f" for version '{version}'" if version else "")
+            + (f"\nAvailable configs in the cache: {available_configs}" if available_configs else "")
+            + (f"\nAvailable versions in the cache: {available_versions}" if version and available_versions else "")
+        )
+    # get most recent
+    cached_directory_path = Path(sorted(cached_directory_paths, key=_get_modification_time)[-1])
+    version, hash = cached_directory_path.parts[-2:]
+    other_configs = [
+        Path(cached_directory_path).parts[-3]
+        for cached_directory_path in glob.glob(os.path.join(cached_datasets_directory_path_root, "*", version, hash))
+        if os.path.isdir(cached_directory_path)
+    ]
+    if not config_name and len(other_configs) > 1:
+        raise ValueError(
```

Comment:
nit: i think we usually use `textwrap.dedent` for multiline error messages like this for readability

Reply:
I just checked and we mostly use `\n` actually - not a big deal imo